feat(plugins): add datadog tracing plugin (#5046)
Kong/docs.konghq.com: app/_hub/kong-inc/datadog-tracing/_index.md

    - name: connect_timeout
      default: 1000
      datatype: number
      description: |
        The timeout, in milliseconds, for the OTLP server connection.
    - name: send_timeout
      default: 5000
      datatype: number
      description: |
        The timeout, in milliseconds, for sending spans to the OTLP server.
    - name: read_timeout
      default: 5000
      datatype: number
      description: |
        The timeout, in milliseconds, for reading the response from the OTLP server.
---

## Usage

{:.note}
> **Note**: The Datadog Tracing plugin only works with {{site.base_gateway}}'s `tracing_instrumentations` configuration enabled. It is recommended that you set the `tracing_instrumentations` configuration to `request`, to avoid impacting performance, and only trace requests.

The Datadog Agent is required to run the Datadog Tracing plugin. The Datadog Agent can be installed on the same machine as {{site.base_gateway}} or on a different machine. The Datadog Tracing plugin will send spans to the Datadog Agent over HTTP.

### Set up {{site.base_gateway}}

Enable the OpenTelemetry tracing capability in {{site.base_gateway}}'s configuration:

> - `tracing_instrumentations = request`: valid values can be found in the [tracing instrumentations](/gateway/latest/reference/configuration/#tracing_instrumentations) section of the configuration documentation.

Reviewer: ```suggestion
- `tracing_instrumentations = request`: Valid values can be found in the [tracing instrumentations](/gateway/latest/reference/configuration/#tracing_instrumentations) section of the configuration documentation.
```

> - `tracing_sampling_rate = 1.0`: tracing instrumentation sampling rate.

Reviewer: ```suggestion
- `tracing_sampling_rate = 1.0`: Tracing instrumentation sampling rate.
```

  The tracer samples a fixed percentage of all spans following the sampling rate.
  This allows you to set the sampling rate to a lower value to reduce the impact of the instrumentation on {{site.base_gateway}}'s proxy performance in production.

### Set up the Datadog Agent

> 

Reviewer: Needs an intro to the section. Something like:

```suggestion

Set up the Datadog Agent. You can use the following example Docker configuration, substituting the `example-api-key` with your own key:

```

```bash

> docker run -d -p 8126:8126 \
>               --cgroupns host \
>               --pid host \
>               -v /var/run/docker.sock:/var/run/docker.sock:ro \
>               -v /proc/:/host/proc/:ro \
>               -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
>               -e DD_API_KEY=<api-key> \
>               -e DD_APM_NON_LOCAL_TRAFFIC=true \
>               -e DD_APM_ENABLED=true \
>               gcr.io/datadoghq/agent:latest

Reviewer: ```suggestion
docker run -d -p 8126:8126 \
  --cgroupns host \
  --pid host \
  -v /var/run/docker.sock:/var/run/docker.sock:ro \
  -v /proc/:/host/proc/:ro \
  -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
  -e DD_API_KEY=<example-api-key> \
  -e DD_APM_NON_LOCAL_TRAFFIC=true \
   -e DD_APM_ENABLED=true \
   gcr.io/datadoghq/agent:latest
```
No need for such deep indents.

```

> For more information about installing the Datadog Agent on Docker, see [Datadog's documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux). If you want to install another container option you can find the instructions [on this list](https://docs.datadoghq.com/containers/).

Reviewer: ```suggestion
For more information about installing the Datadog Agent on Docker, see [Datadog's documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux). If you want to install another container option, you can find the instructions in [Datadog's container list](https://docs.datadoghq.com/containers/).
```
We should use less generic link titles for better SEO and accessibility.

### Configure the Datadog Tracing plugin

Enable the plugin:

```bash
curl -X POST http://<admin-hostname>:8001/plugins \
    --data "name=datadog-tracing"  \
    --data "config.endpoint=http://localhost:8126/v1/traces" \
    --data "config.service_name=kong" \
    --data "config.environment=prod"
```

> ## OpenTelemetry Functionality

Reviewer: ```suggestion
## OpenTelemetry functionality
```

### Instrumentations

{{site.base_gateway}} has a series of built-in tracing instrumentations
which are configured by the `tracing_instrumentations` configuration.

{{site.base_gateway}} creates a top-level span for each request by default when `tracing_instrumentations` is enabled.

The top level span has the following attributes:
- `http.method`: HTTP method
- `http.url`: HTTP URL
- `http.host`: HTTP host
- `http.scheme`: HTTP scheme (http or https)
- `http.flavor`: HTTP version
- `net.peer.ip`: Client IP address

### Propagation

> The OpenTelemetry plugin propagates the following headers:

Reviewer: ```suggestion
The Datadog Tracing plugin propagates the following headers:
```

- `w3c`: [W3C trace context](https://www.w3.org/TR/trace-context/)
- `b3` and `b3-single`: [Zipkin headers](https://github.com/openzipkin/b3-propagation)
- `jaeger`: [Jaeger headers](https://www.jaegertracing.io/docs/client-libraries/#propagation-format)
- `ot`: [OpenTracing headers](https://github.com/opentracing/specification/blob/master/rfc/trace_identifiers.md)
- `Datadog`: [Datadog headers](https://docs.datadoghq.com/tracing/agent/propagation/)

The plugin detects the propagation format from the headers and will use the appropriate format to propagate the span context.
If no appropriate format is found, the plugin will fallback to the default format, which is `Datadog`.

## Troubleshooting

The Datadog Tracing spans are printed to the console when the log level is set to `debug` in the Kong configuration file.

> An example of debug logs output:

Reviewer: ```suggestion
Here's an example of the debug log output:
```

```bash
2022/06/02 15:28:42 [debug] 650#0: *111 [lua] instrumentation.lua:302: runloop_log_after(): [tracing] collected 6 spans:
Span #1 name=GET /wrk duration=1502.994944ms attributes={"http.url":"/wrk","http.method":"GET","http.flavor":1.1,"http.host":"127.0.0.1","http.scheme":"http","net.peer.ip":"172.18.0.1"}
Span #2 name=rewrite phase: datadog-tracing duration=0.391936ms
Span #3 name=router duration=0.013824ms
Span #4 name=access phase: cors duration=1500.824576ms
Span #5 name=cors: heavy works duration=1500.709632ms attributes={"username":"kongers"}
Span #6 name=balancer try #1 duration=0.99328ms attributes={"net.peer.ip":"104.21.11.162","net.peer.port":80}
```

### Known issues

> - Datadog tracing Only supports the HTTP protocols (http/https) of {{site.base_gateway}}.

Reviewer: ```suggestion
- Datadog Tracing only supports the HTTP protocols (`http`/`https`) of {{site.base_gateway}}.
```

> - May impact the performance of {{site.base_gateway}}.

Reviewer: ```suggestion
- This plugin may impact the performance of {{site.base_gateway}}.
```

>   It's recommended to set the sampling rate (`tracing_sampling_rate`)

Reviewer: ```suggestion
  We recommend setting the sampling rate (`tracing_sampling_rate`)
```

>   via Kong configuration file when using the Datadog Tracing plugin.

Reviewer: ```suggestion
  in the {{site.base_gateway}} configuration file when using the Datadog Tracing plugin.
```

- `tracing_instrumentations = request`, Valid values can be found in the [Kong's configuration](/gateway/latest/reference/configuration/#tracing_instrumentations).
- `tracing_sampling_rate = 1.0`: Tracing instrumentation sampling rate.
  Tracer samples a fixed percentage of all spans following the sampling rate.
  Set the sampling rate to a lower value to reduce the impact of the instrumentation on {{site.base_gateway}}'s proxy performance in production.

### Set up a Datadog Agent

```bash
docker run -d -p 8126:8126 \
              --cgroupns host \
              --pid host \
              -v /var/run/docker.sock:/var/run/docker.sock:ro \
              -v /proc/:/host/proc/:ro \
              -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
              -e DD_API_KEY=<api-key> \
              -e DD_APM_NON_LOCAL_TRAFFIC=true \
              -e DD_APM_ENABLED=true \
              gcr.io/datadoghq/agent:latest
```

For more information, see the [Datadog documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux).

### Configure the Datadog Tracing plugin

Enable the plugin:

```bash
curl -X POST http://<admin-hostname>:8001/plugins \

>     --data "name=Datadog-tracing"  \

Reviewer 2: ```suggestion
    --data "name=datadog-tracing"  \
```
